ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala

```diff
@@ -75,25 +82,58 @@ class PreExecutingSubmissionValidatorSpec extends AsyncWordSpec with Matchers wi
     "return a sorted read set with correct fingerprints" in {
       val expectedReadSet =
         TestHelper.allDamlStateKeyTypes.map(key => key -> key.toByteString).toMap
+      val actualInputState =
+        TestHelper.allDamlStateKeyTypes
+          .map(key => key -> ((Some(DamlStateValue.getDefaultInstance), key.toByteString)))
+          .toMap
       val instance = createInstance(expectedReadSet = expectedReadSet)
-      val ledgerStateReader = createLedgerStateReader(expectedReadSet)
+      val ledgerStateReader = createLedgerStateReader(actualInputState)
 
       instance
-        .validate(
-          anEnvelope(expectedReadSet.keySet),
-          aCorrelationId,
-          aParticipantId,
-          ledgerStateReader)
-        .map { actual =>
-          val expectedSortedReadSet = expectedReadSet
-            .map {
-              case (key, fingerprint) =>
-                keySerializationStrategy.serializeStateKey(key) -> fingerprint
-            }
-            .toSeq
-            .sortBy(_._1.asReadOnlyByteBuffer())
-          actual.readSet shouldBe expectedSortedReadSet
-        }
+        .validate(anEnvelope(expectedReadSet.keySet), aParticipantId, ledgerStateReader)
+        .map(verifyReadSet(_, expectedReadSet))
+    }
+
+    "return a read set when the contract keys are inconsistent" in {
+      val contractKeyStateKey = DamlStateKey.newBuilder
+        .setContractKey(
+          DamlContractKey.newBuilder.setTemplateId(Identifier.newBuilder.addName("id")))
+        .build
+      val contractKeyFingerprint = ByteString.copyFromUtf8("contract key fingerprint")
+
+      // At the time of pre-execution, the key points to contract A.
+      val contractIdAStateKey =
```

Comment:
To improve readability, you might want to add constants or helper methods to create these.

Reply:
Done.